### PR TITLE
[FF-A][TPM] Updated TPM Service for Proper State Transitions

### PR DESCRIPTION
**Pull request by Raymond-MS**
## Description

Implemented a translation library for the TPM service to translate the Internal PC CRB to an external TPM type (CRB/FIFO). Updated the TPM service to support going through each phase of the TPM (cmdReady/Start/goIdle). Streamlined the Internal CRB clean-up. Updated the FF-A TPM library to go through the normal state procedure (cmdReady/Start/goIdle).

For details on how to complete these options and their meaning refer to [CONTRIBUTING.md](https://github.com/microsoft/mu/blob/HEAD/CONTRIBUTING.md).

- [x] Impacts functionality?
- [ ] Impacts security?
- [ ] Breaking change?
- [ ] Includes tests?
- [ ] Includes documentation?
- [ ] Backport to release branch?

## How This Was Tested

Verified by running the build with TPM2_ENABLE=TRUE. Verified with the FfaPartitionTestApp.

## Integration Instructions

N/A

### CODE REVIEW

File: ArmPkg/Library/TpmServiceStateTranslationLib/TpmServiceStateTranslationLib.inf [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  SecurityPkg/SecurityPkg.dec
+
+[LibraryClasses]
+  BaseLib
+  BaseMemoryLib
+  IoLib
+  TimerLib
+  DebugLib
+  PlatformFfaInterruptLib
+  ArmSvcLib
```
Comment (kuqin12) [inline ArmPkg/Library/TpmServiceStateTranslationLib/TpmServiceStateTranslationLib.inf]: Do you need all these libraries here?

File: ArmPkg/Include/Library/TpmServiceStateTranslationLib.h [hunk 2]
```diff
@@ -0,0 +1,93 @@
+/** @file
+  Definitions for the TPM Service State Translation Library
```
Comment (kuqin12) [inline ArmPkg/Include/Library/TpmServiceStateTranslationLib.h]: Just curious, why is it "state translation"? It is more than "translation", but also committing the message, right?

Comment (kuqin12) [reply ArmPkg/Include/Library/TpmServiceStateTranslationLib.h]: I think it makes sense to have some more description on the rationale of this library creation. I was going to call it TpmHalLib, but that might be too vague.

File: ArmPkg/Library/TpmServiceLib/TpmServiceLib.c [hunk 3]
```diff
@@ -21,17 +21,39 @@
...
+  NUM_TPM_STATES
+} TpmState;
+
+/* TPM Service Localities */
+typedef enum {
+  TPM_LOCALITY_0 = 0,
+  TPM_LOCALITY_1,
+  TPM_LOCALITY_2,
+  TPM_LOCALITY_3,
+  TPM_LOCALITY_4,
```
Comment (kuqin12) [inline ArmPkg/Library/TpmServiceLib/TpmServiceLib.c]: Hmm, I think this is making the assumption that you are supporting all 5 localities. Do we reference the 5 pages here implicitly?

File: ArmPkg/Library/TpmServiceLib/TpmServiceLib.c [hunk 4]
```diff
@@ -21,17 +21,39 @@
...
+  NUM_TPM_LOCALITIES
```
Comment (kuqin12) [inline ArmPkg/Library/TpmServiceLib/TpmServiceLib.c]: btw, i just noticed that `NUM_LOCALITIES` is defined in the `Tpm20.h`

File: ArmPkg/Library/TpmServiceLib/TpmServiceLib.c [hunk 5]
```diff
@@ -21,17 +21,39 @@
...
-// Default Value - CRB Interface Selected, Only Locality0 Supported
-#define TPM_INTERFACE_ID_DEFAULT  (0x4011)
+#define TPM_CTRL_REQ_CMD_READY_MASK  (0x01)
+#define TPM_CTRL_REQ_GO_IDLE_MASK    (0x02)
+#define TPM_CTRL_STS_IDLE_MASK       (0x02)
+
+#define TPM_LOCALITY_OFFSET  (0x1000)
+
+#define TPM_LOC_STATE_DEFAULT     (0x01)   // Default Value - tpmEstablished
+#define TPM_INTERFACE_ID_DEFAULT  (0x4111) // Default Value - CRB Interface, all localities supported
```
Comment (kuqin12) [inline ArmPkg/Library/TpmServiceLib/TpmServiceLib.c]: Hmm, I think it is might better to use a `PTP_CRB_INTERFACE_IDENTIFIER` struct and populate the sub fields? 0x4111 probably encoded too much implicit information :)

File: ArmPkg/Library/TpmServiceLib/TpmServiceLib.c [hunk 6]
```diff
@@ -21,17 +21,39 @@
...
-#define TPM_LOC_STS_MASK     (0x01)
-#define TPM_CTRL_START_MASK  (0x01)
-
-// Default Value - tpmEstablished
-#define TPM_LOC_STATE_DEFAULT  (0x01)
-// Default Value - CRB Interface Selected, Only Locality0 Supported
-#define TPM_INTERFACE_ID_DEFAULT  (0x4011)
+#define TPM_CTRL_REQ_CMD_READY_MASK  (0x01)
+#define TPM_CTRL_REQ_GO_IDLE_MASK    (0x02)
+#define TPM_CTRL_STS_IDLE_MASK       (0x02)
```
Comment (kuqin12) [inline ArmPkg/Library/TpmServiceLib/TpmServiceLib.c]: Are these

PTP_CRB_CONTROL_AREA_REQUEST_COMMAND_READY
PTP_CRB_CONTROL_AREA_REQUEST_GO_IDLE
PTP_CRB_CONTROL_AREA_STATUS_TPM_IDLE

respectively?

File: ArmPkg/Library/TpmServiceLib/TpmServiceLib.c [hunk 7]
```diff
@@ -75,72 +97,93 @@ ConvertEfiToTpmStatus (
...
-  VOID
-  )
-{
-  /* TODO: If we end up supporting more than Locality0, this will need to clean the correct locality registers */
-  PTP_CRB_REGISTERS_PTR  InternalTpmCrb;
+  /* Set the locality state based on the active locality. */
+  InternalTpmCrb->LocalityState = TPM_LOC_STATE_DEFAULT;
+  switch (ActiveLocality) {
+    case TPM_LOCALITY_0:
+      InternalTpmCrb->LocalityState |= PTP_CRB_LOCALITY_STATE_ACTIVE_LOCALITY_0;
```
Comment (kuqin12) [inline ArmPkg/Library/TpmServiceLib/TpmServiceLib.c:140]: Why is this "bit or"?

File: ArmPkg/Library/TpmServiceLib/TpmServiceLib.c [hunk 8]
```diff
@@ -21,17 +21,39 @@
...
+  TPM_LOCALITY_2,
+  TPM_LOCALITY_3,
+  TPM_LOCALITY_4,
+  NUM_TPM_LOCALITIES
+} TpmLocality;
 
 typedef UINTN TpmStatus;
 
+/* TPM Service Variables */
+static TpmState  CurrentState;
```
Comment (kuqin12) [inline ArmPkg/Library/TpmServiceLib/TpmServiceLib.c]:
```suggestion
STATIC TpmState  mCurrentState;
```

File: ArmPkg/Library/TpmServiceLib/TpmServiceLib.c [hunk 9]
```diff
@@ -21,17 +21,39 @@
...
+static UINT8     ActiveLocality;
```
Comment (kuqin12) [inline ArmPkg/Library/TpmServiceLib/TpmServiceLib.c]:
```suggestion
STATIC UINT8     mActiveLocality;
```

File: ArmPkg/Library/TpmServiceLib/TpmServiceLib.c [hunk 10]
```diff
@@ -158,53 +201,120 @@ HandleCommand (
...
+      if (InternalTpmCrb->CrbControlRequest & TPM_CTRL_REQ_GO_IDLE_MASK) {
+        DEBUG ((DEBUG_INFO, "READY State - Handle TPM Command goIdle Request\n"));
+        Status = TpmSstGoIdle (ActiveLocality);
+        if (Status == EFI_SUCCESS) {
+          CurrentState = TPM_STATE_IDLE;
+        }
+
+        /* Check the cmdReady bit in the CrbControlRequest register, clear it if it has been
+         * set again. */
+      } else if (InternalTpmCrb->CrbControlRequest & TPM_CTRL_REQ_CMD_READY_MASK) {
```
Comment (kuqin12) [inline ArmPkg/Library/TpmServiceLib/TpmServiceLib.c]: For this case, you do not transit state?

File: ArmPkg/Library/TpmServiceStateTranslationLib/TpmServiceStateTranslationLib.inf [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+#
+#  Copyright (c), Microsoft Corporation.
+#
+#  SPDX-License-Identifier: BSD-2-Clause-Patent
+#
+#**/
+
+[Defines]
+  INF_VERSION                    = 1.29
+  BASE_NAME                      = TpmServiceStateTranslationLib
```
Comment (kuqin12) [inline ArmPkg/Library/TpmServiceStateTranslationLib/TpmServiceStateTranslationLib.inf:13]: I think this is the Tis specific instance? If so, you might want to add that to the name. There might be a ptp instance for the same library class.

File: ArmPkg/Library/TpmServiceStateTranslationLib/TpmServiceStateTranslationLib.c [hunk 12]
```diff
@@ -0,0 +1,646 @@
...
+
+#define LOCALITY_OFFSET  (0x1000)
+
+#define DELAY_AMOUNT  (30)
+
+#define PTP_TIMEOUT_MAX  (90000 * 1000) // 90s
+
+/* TPM Service State Translation Library Variables */
+STATIC BOOLEAN  IsCrbInterface;
+STATIC BOOLEAN  IsIdleBypassSupported;
```
Comment (kuqin12) [inline ArmPkg/Library/TpmServiceStateTranslationLib/TpmServiceStateTranslationLib.c]: We normally have a "m" as the prefix of globals.

File: ArmPkg/Library/TpmServiceStateTranslationLib/TpmServiceStateTranslationLib.c [hunk 13]
```diff
@@ -0,0 +1,646 @@
+/** @file
+  Implementation for the TPM Service State Translation Library
+
+  Copyright (c), Microsoft Corporation.
```
Comment (kuqin12) [inline ArmPkg/Library/TpmServiceStateTranslationLib/TpmServiceStateTranslationLib.c:10]: If there is code inherited from open source version, please include the original copyright.

Comment (kuqin12) [reply ArmPkg/Library/TpmServiceStateTranslationLib/TpmServiceStateTranslationLib.c:10]: The same for other files involved in this PR.

Comment (kuqin12) [reply ArmPkg/Library/TpmServiceStateTranslationLib/TpmServiceStateTranslationLib.c:10]: The same for other files involved in this PR.

**Comment by kuqin12** [comment]
I think the change overall makes sense. Two requests would be:
1. Can we add a state machine flowchart? That will help to inspect the edge cases.
2. Add some host-based test for the handlecommand function. (this does not need to come in the same PR)

File: ArmPkg/Include/Library/TpmServiceStateTranslationLib.h (hunk 2 above)
Comment (Raymond-MS) [reply ArmPkg/Include/Library/TpmServiceStateTranslationLib.h]: The library is mainly there to translate the different states the TPM goes through (IDLE, READY, COMPLETE). This is why it exposes GoIdle, CommandReady, and Start. The only thing not really related to states is RequestLocality. I didn't want to call it HAL because it isn't really a hardware abstraction layer. Did you want me to remove "State"? I was thinking of doing so but figured it wouldn't be worth the effort. Up to you, feel free to make the call. :)

File: ArmPkg/Library/TpmServiceLib/TpmServiceLib.c (hunk 6 above)
Comment (Raymond-MS) [reply ArmPkg/Library/TpmServiceLib/TpmServiceLib.c]: Yes, will remove and use those instead. Thanks!

File: ArmPkg/Library/TpmServiceLib/TpmServiceLib.c (hunk 5 above)
Comment (Raymond-MS) [reply ArmPkg/Library/TpmServiceLib/TpmServiceLib.c]: Sure, I can do that. :)

File: ArmPkg/Library/TpmServiceLib/TpmServiceLib.c (hunk 3 above)
Comment (Raymond-MS) [reply ArmPkg/Library/TpmServiceLib/TpmServiceLib.c]: As far as I know, the plan is to support all localities which is why I made this change. I believe so, yes. The localities are sequential.

File: ArmPkg/Library/TpmServiceLib/TpmServiceLib.c (hunk 7 above)
Comment (Raymond-MS) [reply ArmPkg/Library/TpmServiceLib/TpmServiceLib.c:140]: Doing just a "=" would overwrite the default above.

File: ArmPkg/Library/TpmServiceLib/TpmServiceLib.c (hunk 10 above)
Comment (Raymond-MS) [reply ArmPkg/Library/TpmServiceLib/TpmServiceLib.c]: No, we are already in the READY state.

File: ArmPkg/Library/TpmServiceStateTranslationLib/TpmServiceStateTranslationLib.inf (hunk 11 above)
Comment (Raymond-MS) [reply ArmPkg/Library/TpmServiceStateTranslationLib/TpmServiceStateTranslationLib.inf:13]: This is not TIS specific. The library supports both TIS and CRB translation.

File: ArmPkg/Library/TpmServiceStateTranslationLib/TpmServiceStateTranslationLib.inf (hunk 1 above)
Comment (Raymond-MS) [reply ArmPkg/Library/TpmServiceStateTranslationLib/TpmServiceStateTranslationLib.inf]: Probably not, removed.

File: ArmPkg/Library/TpmServiceStateTranslationLib/TpmServiceStateTranslationLib.inf (hunk 11 above)
Comment (kuqin12) [reply ArmPkg/Library/TpmServiceStateTranslationLib/TpmServiceStateTranslationLib.inf:13]: Hmm, do you prefer to carry the extra weight for this service? I guess this library covers the Fifo and Crb case. But that may/will not be all. i.e. it might talk to another fTPM.

Probably want to add some specifier for this instance anyway.

**Review by kuqin12 — APPROVED** [review]

File: ArmPkg/Include/Library/TpmServiceStateTranslationLib.h [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+/**
+  Initiates the transition to the Idle state
+
+  @param  Locality The locality of the TPM to set into Idle
+
+  @retval EFI_SUCCESS  Success
+  @retval EFI_TIMEOUT  Timeout
+
+**/
+EFI_STATUS
```
Comment (apop5) [inline ArmPkg/Include/Library/TpmServiceStateTranslationLib.h:27]: With this being a library interface, should these functions have an EFIAPI calling convention?

File: ArmPkg/Library/TpmServiceLib/TpmServiceLib.c [hunk 15]
```diff
@@ -158,53 +197,120 @@ HandleCommand (
...
-                         &ResponseDataLen,
-                         TpmCommandBuffer
-                         );
+      break;
 
-  /* Copy the response data from the static buffer */
-  CopyMem (InternalTpmCrb->CrbDataBuffer, TpmCommandBuffer, ResponseDataLen);
+    /* The proper state flow should be: IDLE -> READY -> START -> IDLE/READY. */
+    default:
+      break;
```
Comment (apop5) [inline ArmPkg/Library/TpmServiceLib/TpmServiceLib.c:305]: If the state gets out of sorts, should default be used as a way to catch it? Or recover?
i.e. should a debug message be added and/or mCurrentState get reset to something

**Review by apop5 — APPROVED** [review]

File: ArmPkg/Library/TpmServiceLib/TpmServiceLib.c (hunk 15 above)
Comment (Raymond-MS) [reply ArmPkg/Library/TpmServiceLib/TpmServiceLib.c:305]: Yes, you are right. It shouldn't be the case but if for some reason the state gets corrupted, I should handle it gracefully. Good catch. Thanks!

File: ArmPkg/Include/Library/TpmServiceStateTranslationLib.h (hunk 14 above)
Comment (Raymond-MS) [reply ArmPkg/Include/Library/TpmServiceStateTranslationLib.h:27]: As per our discussion, going to leave this as is for now.
